### PR TITLE
Updates to Helix Cluster Map Update tool

### PR DESCRIPTION
**Pull request by pnarayanan**
- Add support for replica count changes for partitions.
- Add an option for force removals of instances and partitions.
- Add an option for validate-only.

### CODE REVIEW

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 1]
```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
...
    */
 
-  private void updateClusterMapInHelixAlt(HelixAdminFactory helixAdminFactory) {
+  private void updateClusterMapInHelix() {
     info("Initializing admins and possibly adding cluster in Helix (if non-existent)");
-    initializeAdminsAndAddCluster(helixAdminFactory);
+    maybeAddCluster();
     info("Initialized");
     info("Populating resources and partitions set");
     info("Populated resources and partitions set");
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: The info is supposed to happen after `populateInstancesAndPartitionsMap()`

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 2]
```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
...
+    int fromIndex = 0;
+    List<Map.Entry<String, Set<String>>> newPartitions = new ArrayList<>(partitionsToInstancesInDc.entrySet());
+    while (fromIndex < newPartitions.size()) {
+      String resourceName = Integer.toString(++maxResource);
+      int toIndex = Math.min(fromIndex + maxPartitionsInOneResource, newPartitions.size());
+      List<Map.Entry<String, Set<String>>> partitionsUnderNextResource = newPartitions.subList(fromIndex, toIndex);
+      fromIndex = toIndex;
+      IdealState idealState = new IdealState(resourceName);
+      idealState.setStateModelDefRef(LeaderStandbySMD.name);
+      info("Adding partitions for instances in " + dcName);
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: use log formatting `info("Adding partitions for instances in {}", dcName);`

Comment (jsjtzyy) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Would be better if `info("Adding partitions for next resource in {}", dcName);`  ?

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 3]
```diff
@@ -629,8 +685,12 @@ private void verifyDataNodeAndDiskEquivalencyInDc(Datacenter dc, String clusterN
       ensureOrThrow(sealedReplicasInClusterMap.equals(sealedReplicasInHelix),
           "Sealed replicas info mismatch for " + "instance " + instanceName);
     }
-    ensureOrThrow(allInstancesInHelix.isEmpty(),
-        "Following instances in Helix not found in the clustermap " + allInstancesInHelix);
+    if (!expectMoreInHelixDuringValidate) {
```
Comment (vgkholla) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: can we track the ones that are expected to be extra? I think that is more fine grained than having one variable which may suppress actual stuff that should not be missing.

Comment (vgkholla) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Same in the if-elses below

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 4]
```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
...
+    int totalInstances = instancesInBoth.size() + instancesInHelix.size() + instancesInStatic.size();
+    for (String instanceName : instancesInBoth) {
+      InstanceConfig instanceConfigInHelix = dcAdmin.getInstanceConfig(clusterName, instanceName);
+      InstanceConfig instanceConfigFromStatic =
+          createInstanceConfigFromStaticInfo(dcToInstanceNameToDataNodeId.get(dcName).get(instanceName),
+              partitionsToInstancesInDc);
+      // Note about the following check: This compares ZNode data for exact equality, which suffices for now.
+      // However, in the future, this needs to discount comparisons for certain things that are dynamically set by
+      // instances (and hence will have outdated information in the static map). Such as
+      // 1. RO/RW status of replicas.
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: I will make this change in a separate PR. Will need to introduce a flag for overriding (as we may want to manually set RO/RW as well by updating these files). Opened https://github.com/linkedin/ambry/issues/923

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 5]
```diff
@@ -88,9 +87,18 @@
   private final Map<String, HelixAdmin> adminForDc = new HashMap<>();
   private final HashMap<String, HashMap<DiskId, SortedSet<Replica>>> instanceToDiskReplicasMap = new HashMap<>();
   private final HashMap<String, HashMap<String, DataNodeId>> dcToInstanceNameToDataNodeId = new HashMap<>();
+  private int maxResource = -1;
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Done.

File: ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java [hunk 6]
```diff
@@ -154,23 +154,58 @@ public void testEverything() throws Exception {
...
       }
     }
-    writeBootstrapOrUpgrade(expectedResourceCount);
+    writeBootstrapOrUpgrade(expectedResourceCount, false);
+
+    // Now, change the replica count for a partition.
+    Partition partition1 = (Partition) testPartitionLayout.getPartitionLayout()
+        .getPartitions(null)
+        .get(RANDOM.nextInt(testPartitionLayout.getPartitionCount()));
+    Partition partition2 = (Partition) testPartitionLayout.getPartitionLayout()
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-tools/src/integration-test/java/com.github.ambry/clustermap/HelixBootstrapUpgradeToolTest.java]: This is relying on `Random.nextInt()` not returning the same value in successive calls (which I think is a fair assumption to make).

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 7]
```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
...
+      // Note about the following check: This compares ZNode data for exact equality, which suffices for now.
+      // However, in the future, this needs to discount comparisons for certain things that are dynamically set by
+      // instances (and hence will have outdated information in the static map). Such as
+      // 1. RO/RW status of replicas.
+      // 2. Replica availability (for which support has not yet been added).
+      // 3. xid (for which support has not yet been added).
+      if (!instanceConfigFromStatic.getRecord().equals(instanceConfigInHelix.getRecord())) {
+        info("Instance {} already present in Helix, but InstanceConfig has changed, updating. Remaining instances: {}",
+            instanceName, --totalInstances);
+        // Continuing on the note above, if there is indeed a change, we must make a call on whether RO/RW, replica
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Opened https://github.com/linkedin/ambry/issues/923, we should do this based on input arguments to the tool to force update or not.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 8]
```diff
@@ -245,131 +283,182 @@ private void populateInstancesAndPartitionsMap() {
...
+    List<String> resourcesInCluster = dcAdmin.getResourcesInCluster(clusterName);
+    for (String resourceName : resourcesInCluster) {
+      boolean resourceModified = false;
+      if (!resourceName.matches("\\d+")) {
+        // there may be other resources created under the cluster (say, for stats) that are not part of the
+        // cluster map. These will be ignored.
+        continue;
+      }
+      maxResource = Math.max(maxResource, Integer.valueOf(resourceName));
+      IdealState resourceIs = dcAdmin.getResourceIdealState(clusterName, resourceName);
```
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Not a typo, `resourceIs` is short for `resourceIdealState`

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: `forceUpdateSealedState` should be used with care and means that regardless of what the node states we should go ahead and make the change. If we don't want a node to invalidate the action done via forced updates, then we should first ensure that the node's configs are set correctly. For example, if we want to force seal a replica on a certain node, then that node's low watermark for unsealing should be lowered.

The point is that this is a flag to add flexibility in how we manipulate the clustermap.

**Comment by pnarayanan** [comment]
@vgkholla @jsjtzyy Comments addressed.

File: ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java [hunk 9]
```diff
@@ -87,12 +87,14 @@
   private final Map<String, HelixAdmin> adminForDc = new HashMap<>();
   private final HashMap<String, HashMap<DiskId, SortedSet<Replica>>> instanceToDiskReplicasMap = new HashMap<>();
   private final HashMap<String, HashMap<String, DataNodeId>> dcToInstanceNameToDataNodeId = new HashMap<>();
```
Comment (jsjtzyy) [inline ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Would be better if we change `HashMap<String, HashMap<DiskId, SortedSet<Replica>>>` to `Map<String, HashMap<DiskId, SortedSet<Replica>>>`;  

Same for `HashMap<String, HashMap<String, DataNodeId>>`

Comment (pnarayanan) [reply ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java]: Addressed this in the latest commit.

**Pull request merged**